src/psyclone/transformations.py

```diff
@@ -563,6 +566,422 @@ def apply(self, node, options=None):
+            if child is not taskloop and not isinstance(child,
+                                                        (Schedule, Loop)):
+                taskloop_vars.merge(VariablesAccessInfo(child))
+        taskloop_signatures = taskloop_vars.all_signatures
+        # Find our parent serial region if it has a barrier
+        parent_single = taskloop.ancestor(OMPSingleDirective)
+        # Cache the parent single for use in later if statements
+        cached_parent_single = parent_single
+        # If the parent single region has a nowait clause it can never
+        # act as the dependency for this taskloop, so we set it to None.
+        # The same behaviour occurs implicitly if the parent
+        # OMPSerialDirective is an OMPMasterDirective. Without a blocking
+        # parent region we can never guarantee synchronicity if dependencies
+        # are outside of the parent region.
+        if parent_single is not None and parent_single.nowait:
+            parent_single = None
+        # Find our parent parallel region
+        parent_parallel = taskloop.ancestor(OMPParallelDirective)
+        # Raise an error if there is no parent_parallel region
+        if parent_parallel is None:
+            fwriter = FortranWriter()
+            raise InternalError(LazyString(lambda: "No parent parallel "
+                                           "directive was found for the "
+                                           "taskloop region: {0}".format(
+                                                fwriter(taskloop).rstrip("\n")
+                                                )))
+
+        for node in node_list:
+            if node.abs_position <= taskloop.abs_position:
+                continue
+            # Ignore any children of the taskloop directive
+            anc = node.ancestor(OMPTaskloopDirective)
+            if anc is taskloop:
+                continue
+            node_vars = None
+            if (isinstance(node, OMPTaskwaitDirective) and
+                    (cached_parent_single is
+                     node.ancestor(OMPSingleDirective)) and
+                    cached_parent_single is not None and
+                    cached_parent_single.nowait is False):
+                # If we find a taskwait barrier inside the same
+                # OMPSingleDirective and that OMPSingleDirective has a no
+                # nowait clause, then it acts as a barrier for dependencies
+                # as well, so we return it
+                return node
+            if not isinstance(node, OMPTaskwaitDirective):
+                # For all our other node types we calculate their own
+                # variable accesses
+                node_vars = VariablesAccessInfo()
+                for child in node.walk(nodes.Node):
+                    if child is not node and not isinstance(child,
+                                                            (Schedule, Loop)):
+                        node_vars.merge(VariablesAccessInfo(child))
```

Comment:
Just a quick comment: creating a new instance of VariablesAccessInfo is not necessary, you can simply do:
```
   child.reference_accesses(node_vars)
````
This will add the new accesses to the existing variable (and saves the overhead of creating a new instance, and then merging them together).

Reply:
Fixed, thanks!